lib/terminus.js

```diff
@@ -1,143 +1,171 @@
 
-  res.statusCode = 200
-  res.setHeader('Content-Type', 'application/json')
-  if (info) {
-    return res.end(JSON.stringify({
-      status: 'ok',
-      info: info
-    }))
+    if (Object.keys(this._options.healthChecks).length > 0) {
+      this._decorateWithHealthCheck()
+    }
+
+    // push the signal into the array
+    // for backwards compatability
+    if (!this._options.signals.includes(this._options.signal)) {
+      this._options.signals.push(this._options.signal)
+    }
+    this._decorateWithSignalHandler()
   }
-  res.end(SUCCESS_RESPONSE)
-}
 
-async function sendFailure (res, options) {
-  const { error, onSendFailureDuringShutdown } = options
+  async _cleanup (signal) {
+    if (!this._state.isShuttingDown) {
+      this._state.isShuttingDown = true
+      try {
+        await this._options.beforeShutdown()
+        await this._asyncServerStop()
+        await this._onSignal()
+        await this._options.onShutdown()
+        this._options.signals.forEach(sig => process.removeListener(sig, () => this._cleanup()))
+        process.kill(process.pid, signal)
+      } catch (error) {
+        this._options.logger('error happened during shutdown', error)
+        process.exit(1)
+      }
+    }
+  }
 
-  if (onSendFailureDuringShutdown) {
-    await onSendFailureDuringShutdown()
+  _sendSuccess (info) {
+    const statusResponse = { status: 'ok' }
+    if (info) {
+      statusResponse.info = info
+    }
+    return statusResponse
   }
-  res.statusCode = 503
-  res.setHeader('Content-Type', 'application/json')
-  if (error) {
-    return res.end(JSON.stringify({
+
+  _sendFailure (error) {
+    return {
       status: 'error',
       error: error
-    }))
+    }
   }
-  res.end(FAILURE_RESPONSE)
-}
 
-const intialState = {
-  isShuttingDown: false
-}
+  _executeHealtCheck (url) {
+    return this._options.healthChecks[url]()
+  }
 
-function noop () {}
+  async _onRequest (req, res, listener) {
+    let statusResponse
 
-function decorateWithHealthCheck (server, state, options) {
-  const { healthChecks, logger, onSendFailureDuringShutdown } = options
+    try {
+      statusResponse = await this.getHealthStatus(req.url)
+    } catch (error) {
+      listener(req, res)
+    }
 
-  server.listeners('request').forEach((listener) => {
-    server.removeListener('request', listener)
-    server.on('request', async (req, res) => {
-      if (healthChecks[req.url]) {
-        if (state.isShuttingDown) {
-          return sendFailure(res, { onSendFailureDuringShutdown })
-        }
-        let info
-        try {
-          info = await healthChecks[req.url]()
-        } catch (error) {
-          logger('healthcheck failed', error)
-          return sendFailure(res, { error: error.causes })
-        }
-        return sendSuccess(res, { info })
-      } else {
-        listener(req, res)
-      }
-    })
-  })
-}
+    res.setHeader('Content-Type', 'application/json')
```

Comment:
this logic would run for all the route handlers, not just for our routes - I guess we don't want to interfere with other routes